Search/ReIndex/Provider/DoctrineOrmProvider.php

```diff
@@ -0,0 +1,122 @@
+
+use Doctrine\ORM\EntityManagerInterface;
+use Massive\Bundle\SearchBundle\Search\ReIndex\ReIndexProviderInterface;
+use Metadata\MetadataFactory;
+
+class DoctrineOrmProvider implements ReIndexProviderInterface
+{
+    /**
+     * @var ClassMetadataFactory
+     */
+    private $entityManager;
+
+    /**
+     * @var MetadataFactory
+     */
+    private $searchMetadataFactory;
+
+    /**
+     * @deprecated BC hack. @see this::provide().
+     *
+     * @var object[]
+     */
+    private $cachedEntities = [];
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        MetadataFactory $searchMetadataFactory
+    ) {
+        $this->entityManager = $entityManager;
+        $this->searchMetadataFactory = $searchMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getClassFqns()
+    {
+        $metadataFactory = $this->entityManager->getMetadataFactory();
+        $classFqns = [];
+
+        foreach ($metadataFactory->getAllMetadata() as $classMetadata) {
+            if (null === $this->searchMetadataFactory->getMetadataForClass($classMetadata->name)) {
+                continue;
+            }
+            $classFqns[] = $classMetadata->name;
+        }
+
+        return $classFqns;
+    }
+
+    /**
+     * BC Note: Previous versions of the MassiveSearchBundle expected a collection of entities to
+     *          be returned from the repository via. a custom repository method. The expected behavior
+     *          now is that a query builder will be passed and NOTHING should be returned.
+     *
+     * {@inheritdoc}
+     */
+    public function provide($classFqn, $offset, $maxResults)
+    {
+        if (!empty($this->cachedEntities)) {
+            return $this->sliceEntities($offset, $maxResults);
+        }
+
+        $repository = $this->entityManager->getRepository($classFqn);
+        $metadata = $this->searchMetadataFactory->getMetadataForClass($classFqn);
+
+        $repositoryMethod = $metadata->getOutsideClassMetadata()->getReindexRepositoryMethod();
+
+        $queryBuilder = $repository->createQueryBuilder('d');
+
+        if ($repositoryMethod) {
+            $result = $repository->$repositoryMethod($queryBuilder);
+
+            if (is_array($result)) {
+                @trigger_error('Reindex repository methods should NOT return anything. Use the passed query builder instead.');
+                $this->cachedEntities = $result;
+
+                return $this->sliceEntities($offset, $maxResults);
+            }
+        }
+
+        $queryBuilder->setFirstResult($offset);
+        $queryBuilder->setMaxResults($maxResults);
+
+        return $queryBuilder->getQuery()->execute();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCount($classFqn)
+    {
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->select('count(a.id)');
+        $queryBuilder->from($classFqn, 'a');
```

Comment:
Isn't it possible to use the query builder returned by the provider? This one should also include the security stuff then, shouldn't it?